### PR TITLE
[benchmark] Add integer parsing benchmarks

### PR DESCRIPTION
**Pull request by PatrickPijnappel**
Adds integer parsing benchmarks for an integer parsing rewrite I have coming up.

### CODE REVIEW

**Comment by PatrickPijnappel** [comment]
@swift-ci please benchmark

**Comment by PatrickPijnappel** [comment]
@swift-ci please benchmark

**Comment by PatrickPijnappel** [comment]
@swift-ci Please smoke test

File: benchmark/single-source/IntegerParsing.swift [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    tags: [.validation, .api],
+    setUpFunction: { blackHole(hexStrings) }),
+  BenchmarkInfo(name: "ParseInt.UncommonRadix",
+    runFunction: run_ParseIntFromUncommonRadix,
+    tags: [.validation, .api],
+    setUpFunction: { blackHole(uncommonRadixStrings) }),
+]
+
+private let values: [Int] = Array(-1000...1000) // Give extra weight to low ints
+  + (0..<2000).map { _ in Int.random(in: .min ... .max) }
```
Comment (stephentyrone) [inline benchmark/single-source/IntegerParsing.swift]: It may be worth actually splitting these into two separate tests for each radix.

**Comment by stephentyrone** [comment]
(CC @milseman who may also be interested in this)

File: benchmark/single-source/IntegerParsing.swift [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+@inline(never)
+public func run_ParseIntFromDecimal(N: Int) {
+  var result = 0
+  for _ in 0..<N {
+    for string in decimalStrings {
+      result &+= Int(string, radix: 10)!
+    }
+  }
+  blackHole(result)
```
Comment (milseman) [inline benchmark/single-source/IntegerParsing.swift]: Can you add a CheckResults for `N * your-expected-total`?

**Review by milseman — APPROVED** [review]
Overall LGTM, would be nice to have a `CheckResults` call at the end of the loop as a sanity check.

Looking forwards to integer parsing improvements!

**Comment by PatrickPijnappel** [comment]
@stephentyrone @milseman Done, thanks! Also it probably makes more sense to fix the large values to 64 bits.

**Comment by PatrickPijnappel** [comment]
@swift-ci Please benchmark

**Comment by milseman** [comment]
@swift-ci please test

**Comment by PatrickPijnappel** [comment]
@swift-ci Please benchmark

**Comment by PatrickPijnappel** [comment]
@swift-ci Please smoke test

**Comment by PatrickPijnappel** [comment]
@swift-ci Please smoke test

**Comment by PatrickPijnappel** [comment]
@swift-ci Please benchmark

**Comment by PatrickPijnappel** [comment]
@milseman About ~30% of the benchmark is retain/release within the run method loop, do you know any way to eliminately that? I've been trying a few different things but no luck so far

**Comment by milseman** [comment]
Do you know what is being retained? Could you post more information or a profile?

It's possibly a retain happening inside the integer initializer. These initializers are some of the worst code in the standard library; I've been wanting to rewrite them for a while but it keeps getting put off as being just barely under my prioritization threshold for the next release.

If you're planning on tackling that next, that's really exciting! Would you like to be assigned https://bugs.swift.org/browse/SR-7556 ?

**Comment by PatrickPijnappel** [comment]
@milseman They seem to be outside the initializer so I was assuming the input strings, but since it's specialized you might be right and it might just be a retain that belongs inside it: https://cl.ly/8443f446de87

Yeah! I'm already quite far with it. Just assigned myself to the issue. I've also been looking at integer > string which is currently still in C++ but I have a Swift version now that's quite a bit faster :)

**Comment by milseman** [comment]
Those looks like issues with the current implementation, hopefully yours won't have them!

@swift-ci please test

**Comment by milseman** [comment]
@swift-ci please benchmark

**Comment by milseman** [comment]
@PatrickPijnappel I'm ready to merge if you are

**Comment by PatrickPijnappel** [comment]
Ok!

**Pull request merged**